proc_fdinfo.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package procfs
+
+import (
+	"bufio"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strings"
+)
+
+// ProcFDInfo contains represents file descriptor information.
+type ProcFDInfo struct {
+	// File descriptor
+	FD string
+	// File offset
+	Pos string
+	// File access mode and status flags
+	Flags string
+	// Mount point ID
+	MntID string
+	// List of inotify lines (structed) in the fdinfo file (kernel 3.8+ only)
+	InotifyInfos []InotifyInfo
+}
+
+// FDInfo constructor. On kernels older than 3.8, InotifyInfos will always be empty.
+func (p Proc) FDInfo(fd string) (*ProcFDInfo, error) {
+	f, err := os.Open(p.path("fdinfo", fd))
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	fdinfo, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, fmt.Errorf("could not read %s: %s", f.Name(), err)
+	}
+
+	var text, pos, flags, mntid string
+	var inotify []InotifyInfo
+	rPos := regexp.MustCompile(`^pos:\s+(\d+)$`)
+	rFlags := regexp.MustCompile(`^flags:\s+(\d+)$`)
+	rMntID := regexp.MustCompile(`^mnt_id:\s+(\d+)$`)
+	rInotify := regexp.MustCompile(`^inotify`)
```

Comment:
These regexps should be moved to top level package variables so that they don't need to be compiled for each function call.

Reply:
Sorry, I'm not yet familiar on how this is set. Do I just declare these at the top of the same file, after the imports?

Reply:
Yes, you can put them in a [`var ()` block](https://golang.org/doc/effective_go.html#variables) outside the function.  The location is mostly style preference, but either after the imports or just before the function where they are used are fine.